database-plugins/src/main/java/io/cdap/plugin/db/connector/DBConnector.java

```diff
@@ -0,0 +1,328 @@
+        }
+        count++;
+      }
+    }
+    return browseDetailBuilder.setTotalCount(count).build();
+  }
+
+  private BrowseDetail getTableDetail(Connection connection, @Nullable String schema, String table)
+    throws SQLException {
+    BrowseDetail.Builder browseDetailBuilder = BrowseDetail.builder();
+    DatabaseMetaData metaData = connection.getMetaData();
+    try (ResultSet resultSet = metaData.getTables(connection.getCatalog(), schema, table, SUPPORTED_TYPES)) {
+      if (resultSet.next()) {
+        String name = resultSet.getString(RESULTSET_COLUMN_TABLE_NAME);
+        browseDetailBuilder.addEntity(BrowseEntity
+                                        .builder(name, schema == null ? "/" + name : "/" + schema + "/" + name,
+                                                 resultSet.getString(RESULTSET_COLUMN_TABLE_TYPE)).canSample(true).build());
+      } else {
+        throw new IllegalArgumentException(String.format("Cannot find table : %s.%s.", schema, table));
+      }
+    }
+    return browseDetailBuilder.setTotalCount(1).build();
+  }
+
+  private BrowseDetail listTables(Connection connection, @Nullable String schema,
+                                  @Nullable Integer limit) throws SQLException {
+    BrowseDetail.Builder browseDetailBuilder = BrowseDetail.builder();
+    int count = 0;
+    DatabaseMetaData metaData = connection.getMetaData();
+    // make sure schema exists
+    if (schema != null) {
+      // NOTE Oracle schema name is case sensitive here
+      try (ResultSet schemas = metaData.getSchemas(connection.getCatalog(), schema)) {
+        if (!schemas.next()) {
+          throw new IllegalArgumentException(String.format("Schema '%s' does not exist.", schema));
+        }
+      }
+    }
+    try (ResultSet resultSet = metaData.getTables(connection.getCatalog(), schema, null, SUPPORTED_TYPES)) {
+      int countLimit = limit == null || limit <= 0 ? Integer.MAX_VALUE : limit;
+      while (resultSet.next()) {
+        String name = resultSet.getString(RESULTSET_COLUMN_TABLE_NAME);
+        if (count < countLimit) {
+          browseDetailBuilder.addEntity(BrowseEntity
+                                          .builder(name, schema == null ? "/" + name : "/" + schema + "/" + name,
+                                                   resultSet.getString(RESULTSET_COLUMN_TABLE_TYPE)).canSample(true).build());
+        }
+        count++;
+      }
+    }
+    return browseDetailBuilder.setTotalCount(count).build();
+  }
+
+  private List<StructuredRecord> getTableData(Connection connection, @Nullable String schema,
+    String table, int limit) throws SQLException {
+    String productName = connection.getMetaData().getDatabaseProductName();
+    String query = DBDifferenceUtils.getTableQueryWithLimit(productName, schema, table, limit);
+    return parseResultSet(connection.createStatement().executeQuery(query));
```

Comment:
here when you create a statement, you can use `statement.setFetchSize(limit)`, this way the result set will only retrieve that limit number of data in a round trip even if the query is `SELECT * FROM TABLE`, after you getting the limit, you can then close the resultset and statement.

Reply:
This way you can get rid of the limit hack for specific data types

Reply:
Also this resultset and statement needs to be closed